File: solutions/exercise-7/src/main/java/io/miragon/training/adapter/outbound/cibseven/MembershipProcessAdapter.java
package io.miragon.training.adapter.outbound.cibseven;

import io.miragon.training.application.port.outbound.MembershipProcess;
import io.miragon.training.domain.Membership;
import io.miragon.training.domain.MembershipId;
import org.cibseven.bpm.engine.RuntimeService;
import org.springframework.stereotype.Component;

import java.util.Map;

@Component
public class MembershipProcessAdapter implements MembershipProcess {

    private final RuntimeService runtimeService;

    public MembershipProcessAdapter(RuntimeService runtimeService) {
        this.runtimeService = runtimeService;
    }

    @Override
    public void startProcess(Membership membership) {
        runtimeService.createMessageCorrelation("Message_SubscriptionRequested")
                .setVariables(Map.of(
                        "membershipId", membership.id().value().toString(),
                        "email", membership.email().value(),
                        "name", membership.name().value(),
                        "age", membership.age().value()
                ))
                .correlateStartMessage();
    }

    @Override
    public void rejectMembership(MembershipId membershipId) {
        runtimeService.createMessageCorrelation("Message_ConfirmationRejected")
                .processInstanceVariableEquals("membershipId", membershipId.value().toString())
                .correlate();
    }
}

File: solutions/exercise-4/src/main/kotlin/io/miragon/training/adapter/outbound/cibseven/MembershipProcessAdapter.kt
package io.miragon.training.adapter.outbound.cibseven

import io.miragon.training.adapter.process.MembershipProcessApi
import io.miragon.training.application.port.outbound.MembershipProcess
import io.miragon.training.domain.Membership
import io.miragon.training.domain.MembershipId
import org.cibseven.bpm.engine.RuntimeService
import org.springframework.stereotype.Component

@Component
class MembershipProcessAdapter(
    private val runtimeService: RuntimeService,
) : MembershipProcess {

    override fun startProcess(membership: Membership) {
        runtimeService.startProcessInstanceByKey(
            MembershipProcessApi.PROCESS_KEY,
            mapOf(
                MembershipProcessApi.Variables.MEMBERSHIP_ID to membership.id.value.toString(),
                MembershipProcessApi.Variables.EMAIL to membership.email.value,
                MembershipProcessApi.Variables.NAME to membership.name.value,
                MembershipProcessApi.Variables.AGE to membership.age.value,
            )
        )
    }

    override fun rejectMembership(membershipId: MembershipId) {
        runtimeService.createMessageCorrelation(MembershipProcessApi.Messages.MESSAGE_CONFIRMATION_REJECTED)
            .processInstanceVariableEquals(MembershipProcessApi.Variables.MEMBERSHIP_ID, membershipId.value.toString())
            .correlate()
    }
}
